weblate/trans/models/component.py

```diff
@@ -1152,17 +1152,24 @@ def get_export_url(self):
         return self.git_export
 
     def get_repoweb_link(
-        self, filename: str, line: str, template: Optional[str] = None
+        self, filename: str, line: str, template: Optional[str] = None, user=None
     ):
         """Generate link to source code browser for given file and line.
 
         For linked repositories, it is possible to override the linked repository path
         here.
         """
         if not template:
-            template = self.repoweb
+            if self.repoweb:
+                template = self.repoweb
+            elif user and user.has_perm("vcs.view", self):
+                template = "{}/blob/{}/{}/#{}".format(
+                    self.get_repo_url, self.get_repo_branch, filename, line
+                )
```

Comment:
This should to be handled in a separate method, parse the repo URL, do the detection and return template. To avoid issues with using template syntax in the filename/branch name, it should return the actual template, which will be filled in by filename/line later via `render_template` call, so use `{{branch}} {{filename}} {{line}}` instead of the actual values.

Reply:
I just noticed your questions in #7865, I will provide more detailed answer there...